scripts/scratch_ingest/expression_writer.py

```diff
@@ -0,0 +1,416 @@
+
+        :param line: (String) single line from matrix file
+        :return: (List): values as integers
+        """
+        gene_idx, barcode_idx, raw_exp = line.rstrip().split(' ')
+        return [int(gene_idx), int(barcode_idx), round(float(raw_exp), self.precision)]
+
+    def process_dense_data(self, cluster_cells, cluster_name, data_dir):
+        """
+        Main handler to read dense matrix data and process entries at the gene level
+
+        :param cluster_cells: (List) cell names from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir: (String) name out output dir
+        """
+        pool = multiprocessing.Pool(self.num_cores)
+        slice_indexes = self.get_file_seek_points()
+        matrix_file = ExpressionWriter.open_file(self.matrix_file_path)
+        header = matrix_file.readline().rstrip()
+        values = re.split(COMMA_OR_TAB, header)
+        cells = values[1:]
+        processor = partial(
+            self.read_dense_matrix_slice, cells, cluster_cells, cluster_name, data_dir
+        )
+        pool.map(processor, slice_indexes)
+
+    def read_dense_matrix_slice(self, matrix_cells, cluster_cells, cluster_name, data_dir, indexes):
+        """
+        Read a dense matrix using start/stop indexes and create to individual gene-level files
+        :param matrix_cells: (List) cell names from matrix file
+        :param cluster_cells: (List) cell names from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir: (String) name out output dir
+        :param indexes: (List) start/stop index points to read from/to
+        """
+        start_pos, end_pos = indexes
+        print(f"reading {self.matrix_file_path} at index {start_pos}:{end_pos}")
+        with ExpressionWriter.open_file(self.matrix_file_path) as matrix_file:
+            current_pos = start_pos
+            matrix_file.seek(current_pos)
+            while current_pos < end_pos:
+                line = matrix_file.readline()
+                self.process_dense_line(line, matrix_cells, cluster_cells, cluster_name, data_dir)
+                current_pos += len(line)
+
+    def process_dense_line(self, line, matrix_cells, cluster_cells, cluster_name, data_dir):
+        """
+        Process a single line from a dense matrix and write gene-level data
+
+        :param matrix_cells: (List) cells from header line of dense matrix
+        :param cluster_cells: (List) cell names from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir (String) name out output dir
+        :param line: (String) single line from dense matrix
+        """
+        clean_line = line.rstrip().replace('"','')
+        raw_vals = re.split(COMMA_OR_TAB, clean_line)
+        gene_name = raw_vals.pop(0)
+        exp_vals = [round(float(val), self.precision) if float(val) != 0.0 else 0 for val in raw_vals]
```

Comment:
Per other comment re `round_exp`:
```suggestion
        exp_vals = [round_exp(val, self.precision) if float(val) != 0.0 else 0 for val in raw_vals]
```

Reply:
[36e7b90](https://github.com/broadinstitute/scp-ingest-pipeline/pull/270/commits/36e7b906849129a2177b865d7705ef033995fdbe)